src/util/backend.js

```diff
@@ -0,0 +1,313 @@
+    this.maximum = new BigNumber(opts.maximum)
+    this.minimum = new BigNumber(opts.minimum)
+    this.cache = {}
+    this.key = opts.key || ''
+    if (!this.key.match(KEY_REGEX)) {
+      throw new Error('invalid key: ' + this.key)
+    }
+
+    // optional, for stateful plugin only
+    this.store = store
+    this.writeQueue = Promise.resolve()
+
+    // TODO: disable balance? (not needed for client plugin)
+    this.balance_if = new BigNumber(0)
+    this.balance_i = new BigNumber(0)
+    this.balance_o = new BigNumber(0)
+    this.balance_of = new BigNumber(0)
+  }
+
+  async connect () {
+    if (this.connected) return
+
+    if (this.store) {
+      this.maximum = new BigNumber(await this.store.get(this.key + ':tl:maximum') || this.maximum)
+      this.minimum = new BigNumber(await this.store.get(this.key + ':tl:minimum') || this.minimum)
+      this.balance_if = new BigNumber(await this.store.get(this.key + ':tl:balance:if') || 0)
+      this.balance_of = new BigNumber(await this.store.get(this.key + ':tl:balance:of') || 0)
+      this.balance_i = new BigNumber(this.balance_if)
+      this.balance_o = new BigNumber(this.balance_of)
+    }
+
+    this.connected = true
+  }
+
+  async setMaximum (n) {
+    await this.connect()
+    this.maximum = new BigNumber(n)
+
+    if (this.store) {
+      this.writeQueue = this.writeQueue
+        .then(() => {
+          return this.store.put(this.key + ':tl:maximum', this.maximum.toString())
+        })
+
+      await this.writeQueue
+    }
+  }
+
+  async setMinimum (n) {
+    await this.connect()
+    this.minimum = new BigNumber(n)
+
+    if (this.store) {
+      this.writeQueue = this.writeQueue
+        .then(() => {
+          return this.store.put(this.key + ':tl:minimum', this.minimum.toString())
+        })
+
+      await this.writeQueue
+    }
+  }
+
+  async getMaximum () {
+    await this.connect()
+    return this.maximum.toString()
+  }
+
+  async getMinimum () {
+    await this.connect()
+    return this.minimum.toString()
+  }
+
+  async getBalance () {
+    await this.connect()
+    return this.balance_if.sub(this.balance_of).toString()
+  }
+
+  async getIncomingFulfilled () {
+    await this.connect()
+    return this.balance_if.toString()
+  }
+
+  async getOutgoingFulfilled () {
+    await this.connect()
+    return this.balance_of.toString()
+  }
+
+  async getIncomingFulfilledAndPrepared () {
+    await this.connect()
+    return this.balance_i.toString()
+  }
+
+  async getOutgoingFulfilledAndPrepared () {
+    await this.connect()
+    return this.balance_o.toString()
+  }
+
+  async get (id) {
+    await this.connect()
+    // TODO: errors
+    // - what if the transfer doesn't exist?
+    return this.cache[id] ||
+      (this.store && (await this.store.get(this.key + ':tl:transfer:' + id)))
+  }
+
+  async prepare (transfer, isIncoming) {
+    await this.connect()
+    // TODO: should direction be a boolean isIncoming?
+    // TODO: errors
```

Comment:
These are all the outstanding questions about the PluginBackend spec. Finalizing the spec is one of the currently outstanding tasks for this PR.